src/opm/output/eclipse/Tables.cpp

```diff
@@ -41,8 +48,100 @@
 #include <array>
 #include <cstddef>
 #include <iterator>
+#include <numeric>
+#include <utility>
 #include <vector>
 
+namespace {
+    /// Create linearised, padded TAB vector entries for a collection of
+    /// tabulated saturation functions corresponding to a single input
+    /// keyword.
+    ///
+    /// \tparam BuildDependent Callable entity that extracts the
+    ///    independent and primary dependent variates of a single
+    ///    saturation function table into a linearised output table.
+    ///    Must implement a function call operator of the form
+    ///    \code
+    ///       std::size_t
+    ///       operator()(const std::size_t      tableID,
+    ///                  const std::size_t      primID,
+    ///                  LinearisedOutputTable& table);
+    ///    \endcode
+    ///    that will assign the independent variate of the sub-table
+    ///    primID within the table identified as tableID to column zero
+    ///    of 'table' and all dependent variates to columns one &c of
+    ///    'table'.  The function call operator must return the number
+    ///    of active (used) rows within the sub-table through its return
+    ///    value.
+    ///
+    /// \param[in] numTab Number of tables in this table collection.
+    ///
+    /// \param[in] numPrim Number of primary look-up keys for each table.
+    ///    Mostly relevant (i.e., greater than one) for miscible oil
+    ///    ("PVTO") or miscible gas ("PVTG") tables and typically
+    ///    corresponds to the number of Rs/Rv entries from the TABDIMS
+    ///    keyword.
+    ///
+    /// \param[in] numRows Number of rows to allocate for each padded
+    ///    output table.
+    ///
+    /// \param[in] numDep Number of dependent variates (columns) in each
+    ///    table of this collection of input tables.  Total number of
+    ///    columns in the result vector will be 1 + 2*numDep to account
+    ///    for the independent variate, the dependent variates and the
+    ///    derivatives of the dependent variates with respect to the
+    ///    independent variate.
+    ///
+    /// \param[in] buildDeps Function object that implements the
```

Comment:
If the `buildDeps` argument actually is a callable it could maybe be of type `std::function` instead of being a template parameter - the use of a template parameter makes the code more difficult to comprehend.

Reply:
> If the `buildDeps` argument actually is a callable it could maybe be of type `std::function` instead of being a template parameter

I tried to write `createPropfuncTable()` a bit like an STL algorithm.  Of course it doesn't work directly on iterators that demarcate a range, but the way of passing the loop body as a template parameter was inspired by the STL.  That said, even if using `std::function` will add a little bit of cost&mdash;like reducing the compiler's opportunity for inlining operations and potentially copying the table data&mdash;the resulting code is potentially more familiar to other developers so switching is certainly possible.

---

## Relevant File Sizes (Bytes) ##

As an experiment, and to try to assess the cost of using `std::function` over template, I compiled both the file separately and all of `libopmcommon` with both alternatives.

- `CMakeFiles/src/opm/output/eclipse/Tables.cpp.o`

  Configuration | `std::function` | template
  -----------------|---------------------|-------------
  Debug | 5,844,928 | 5,708,920
  Release | 98,712 | 78,528

- `lib/libopmcommon.a`

  Configuration | `std::function` | template
  -----------------|---------------------|-------------
  Debug | 396,696,656 | 396,559,704
  Release | 12,577,936 | 12,557,752

This suggests that using `std::function` does not introduce inordinate overhead.  I'll switch to using `std::function`.  Thanks for the suggestion.

Reply:
>  I'll switch to using std::function. Thanks for the suggestion.

Good - thank you!